### PR TITLE
Moving ASIC constraints to the 'constraints' group

### PR DESCRIPTION
**Pull request by aolofsson**
- Make area constraints general (applies to other things as well, not just asic)
- Making maxlayer/minlayer per net specific with wildcard (rather than asic global)
- Renaming diearea to outline (diearea is too asic specific)
- Adding uselambda option (to turn off pdk multiplier for those that want it)
- Adding more routing constraints (for analog and pcb routing mostly)
- Adding more pin constraints to gently drive floorplan (side + order)
- Fixing relax option to indicate that it applies to all tools (not just RTL linting)
- Adding partname to placement constraints (for physical only cells)
- Adding global min/max routing layer on a per pdk/stackup basis (can be overridden in constraints)
- Adding missing unit field for physical dimensions (mandatory)

### CODE REVIEW

File: siliconcompiler/schema/schema_cfg.py [hunk 1]
```diff
@@ -3557,6 +3440,129 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+    # NETS
+    scparam(cfg, ['constraint', 'net', name, 'maxlength'],
+            sctype='float',
+            shorthelp="Constraint: Net max length",
+            switch="-constraint_net_maxlength 'name <float>'",
+            example=[
+                "cli: -constraint_net_maxlength 'nreset 1000'",
+                "api: chip.set('constraint', 'net', 'nreset', 'maxlength', '1000')"],
+            schelp="""
+            Maximum total length of a net.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3452]: It's removed from `asic,maxlength` but added back here?. I suppose you could also apply this to {*}, so the help should be updated accordingly.

File: siliconcompiler/schema/schema_cfg.py [hunk 2]
```diff
@@ -3557,6 +3440,129 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            shorthelp="Constraint: Net minimum routing layer",
+            switch="-constraint_net_minlayer 'name <str>'",
+            example=[
+                "cli: -constraint_net_minlayer 'nreset m1'",
+                "api: chip.set('constraint', 'net', 'nreset', 'minlayer', 'm1')"],
+            schelp="""
+            Minimum metal layer to be used for automated place and route
+            specified on a per net basis. Metal names should either be the PDK
+            specific metal stack name or an integer with '1' being the lowest
+            routing layer.Wildcards ('*') can be used for net names.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3477]: layer.Wildcards -> layer. Wildcards

File: siliconcompiler/schema/schema_cfg.py [hunk 3]
```diff
@@ -3557,6 +3440,129 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            shorthelp="Constraint: Net maximum routing layer",
+            switch="-constraint_net_maxlayer 'name <str>'",
+            example=[
+                "cli: -constraint_net_maxlayer 'nreset m1'",
+                "api: chip.set('constraint', 'net', 'nreset', 'maxlayer', 'm1')"],
+            schelp="""
+            Maximum metal layer to be used for automated place and route
+            specified on a per net basis. Metal names should either be the PDK
+            specific metal stack name or an integer with '1' being the lowest
+            routing layer.Wildcards ('*') can be used for net names.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3490]: layer.Wildcards -> layer. Wildcards

File: siliconcompiler/schema/schema_cfg.py [hunk 4]
```diff
@@ -3557,6 +3440,129 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            switch="-constraint_coremargin <float>",
+            example= ["cli: -constraint_coremargin 1",
+                      "api: chip.set('constraint', 'coremargin', '1')"],
+            schelp="""
+            Halo/margin between the outline and core area for fully
+            automated layout sizing and floorplanning.
+            Corearea outline length = outline length * coremargin,
+            where coremaging is a number between 1 (no halo) and 0
+            (no core area). This number is ignored if corearea is
+            supplied.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3536]: Correct this to be coremargin in um again. If you want to keep the number is ignored, we can set the defvalue to 0.0

File: siliconcompiler/schema/schema_cfg.py [hunk 5]
```diff
@@ -3557,6 +3440,129 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            shorthelp="Constraint: Layout density",
+            switch="-constraint_density <float>",
+            example= ["cli: -constraint_density 30",
+                      "api: chip.set('constraint', 'density', '30')"],
+            schelp="""
+            Target density based on the total design cells area reported
+            after synthesis/elaboration. This number is used when no outline
+            or floorplan is supplied. Any number between 1 and 100 is legal,
+            but values above 50 may fail due to area/congestion issues during
+            apr.""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3550]: maybe spell out apr?

File: siliconcompiler/tools/openroad/openroad.py [hunk 6]
```diff
@@ -196,10 +196,10 @@ def setup(chip, mode='batch'):
...
                             ('grt_allow_congestion', 'False'),
                             ('grt_allow_overflow', 'False'),
-                            ('grt_signal_min_layer', chip.get('asic', 'minlayer')),
-                            ('grt_signal_max_layer', chip.get('asic', 'maxlayer')),
-                            ('grt_clock_min_layer', chip.get('asic', 'minlayer')),
-                            ('grt_clock_max_layer', chip.get('asic', 'maxlayer')),
+                            ('grt_signal_min_layer', chip.get('constraint', 'net', '*', 'minlayer')),
+                            ('grt_signal_max_layer', chip.get('constraint', 'net', '*', 'maxlayer')),
+                            ('grt_clock_min_layer', chip.get('constraint', 'net', '*', 'minlayer')),
+                            ('grt_clock_max_layer', chip.get('constraint', 'net', '*', 'maxlayer')),
```
Comment (gadfort) [inline siliconcompiler/tools/openroad/openroad.py:202]: we should probably check if these are set before calling them and add a requirement for them for OpenROAD.

**Review by gadfort — CHANGES_REQUESTED** [review]
Just need to fix coremargin and maybe add some checks in OpenROAD for the new constraints since they are becoming a requirement for the tool

File: siliconcompiler/schema/schema_cfg.py (hunk 4 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:3536]: fixed

File: siliconcompiler/tools/openroad/openroad.py (hunk 6 above)
Comment (aolofsson) [reply siliconcompiler/tools/openroad/openroad.py:202]: Thought about it some more, and I am not quite there with the global constraints yet....needs more consideration.

**Review by gadfort — CHANGES_REQUESTED** [review]

File: examples/gcd/run.sh [hunk 7]
```diff
@@ -12,10 +12,10 @@ sc -design gcd \
    -package_description "GCD test package" \
    -package_license "MIT" \
    -target "freepdk45_demo" \
-   -asic_diearea "(0,0)" \
-   -asic_diearea "(100.13,100.8)" \
-   -asic_corearea "(10.07,11.2)" \
-   -asic_corearea "(90.25,91)" \
+   -constraint_outline "(0,0)" \
+   -constraint_outline "(100.13,100.8)" \
```
Comment (WRansohoff) [inline examples/gcd/run.sh:16]: Not blocking, but is it possible to pass in both rectangle corners with one `-constraint_outline` command-line option?

File: siliconcompiler/schema/schema_cfg.py [hunk 8]
```diff
@@ -2124,6 +2150,17 @@ def schema_option(cfg):
             schelp="""
             Target PDK used during compilation.""")
 
+    scparam(cfg, ['option','uselambda'],
```
Comment (WRansohoff) [inline siliconcompiler/schema/schema_cfg.py:2153]: May be a little ambiguous, especially if the 'lambda' value might end up being a sort of coarse measurement?

**Review by WRansohoff — APPROVED** [review]
LGTM, just a couple of small questions

**Review by gadfort — APPROVED** [review]
lgtm

File: siliconcompiler/schema/schema_cfg.py [hunk 9]
```diff
@@ -3557,6 +3484,209 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            switch="-constraint_net_maxlength 'name <float>'",
+            example=[
+                "cli: -constraint_net_maxlength 'nreset 1000'",
+                "api: chip.set('constraint', 'net', 'nreset', 'maxlength', '1000')"],
+            schelp="""
+            Maximum total length of a net. Wildcards ('*') can be used for net names.""")
+
+    scparam(cfg, ['constraint', 'net', name, 'maxresistance'],
+            sctype='float',
+            shorthelp="Constraint: Net max resistasnce",
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:3566]: done

File: siliconcompiler/schema/schema_cfg.py [hunk 10]
```diff
@@ -3557,6 +3484,209 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+                "api: chip.set('constraint', 'net', 'nreset', 'maxlayer', 'm1')"],
+            schelp="""
+            Maximum metal layer to be used for automated place and route
+            specified on a per net basis. Metal names should either be the PDK
+            specific metal stack name or an integer with '1' being the lowest
+            routing layer. Wildcards ('*') can be used for net names.""")
+
+
+    scparam(cfg, ['constraint', 'net', name, 'shielding'],
+            sctype='bool',
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py]: implemented

File: examples/gcd/run.sh (hunk 7 above)
Comment (aolofsson) [reply examples/gcd/run.sh:16]: not as of now, this limitation applies to all lists and cmdline currently.

File: siliconcompiler/schema/schema_cfg.py [hunk 11]
```diff
@@ -3497,10 +3430,35 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            sctype='(float,float)',
+            unit='um',
+            shorthelp="Constraint: Component halo",
+            switch="-constraint_component_halo 'inst <(float,float)>'",
+            example=[
+                "cli: -constraint_component_halo 'i0 (1,1)'",
+                "api: chip.set('constraint', 'component', 'i0', 'halo', (1,1))"],
+            schelp="""
+            Placement keepout halo around the named component, specified as a
+            (horizontal, vertical) tuple represented in micros on lambda units.
```
Comment (nmoroze) [inline siliconcompiler/schema/schema_cfg.py:3455]: typo: "microns or"

File: siliconcompiler/schema/schema_cfg.py [hunk 12]
```diff
@@ -3543,7 +3504,8 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
             guidelines. The 'z' coordinate shall be set to 0 for planar components
             with only (x,y) coordinates. Discretized systems like 3D chips with
             pins on to and bottom may choose to discretize the top and bottom
-            layer as 0,1 or use absolute coordinates.""")
+            layer as 0,1 or use absolute coordinates. Values are specified based
+            on """)
```
Comment (nmoroze) [inline siliconcompiler/schema/schema_cfg.py:3508]: meant to say more?

File: siliconcompiler/schema/schema_cfg.py [hunk 13]
```diff
@@ -3557,6 +3519,217 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+            switch="-constraint_pin_side 'name <int>'",
+            example=[
+                "cli: -constraint_pin_side 'nreset 1'",
+                "api: chip.set('constraint', 'pin', 'nreset', 'side', 1)"],
+            schelp="""
+            Side of block where the named pin should be placed. Sides are
+            enumerated as an integers with '1' being the lower left side,
+            and incrementing the side with every turn in a clock wise
+            fashion. In case of conflict between 'lower' and and 'left',
+            'left' has precedence. 'The side option and order option are
```
Comment (nmoroze) [inline siliconcompiler/schema/schema_cfg.py:3534]: typo: extra '

**Review by nmoroze — APPROVED** [review]
Looks good, just left a few comments where I noticed typos.

File: siliconcompiler/schema/schema_cfg.py [hunk 14]
```diff
@@ -3557,6 +3484,209 @@ def schema_constraint(cfg, scenario='default', name = 'default'):
...
+                "api: chip.set('constraint', 'pin', 'nreset', 'order', 1)"],
+            schelp="""
+            The relative position of the named pin in a vector of pins
+            on the side specified by the 'side' option. Pin order counting
+            is done clockwise.""")
+
+    # NETS
+    scparam(cfg, ['constraint', 'net', name, 'maxlength'],
+            sctype='float',
+            shorthelp="Constraint: Net max length",
```
[the remark this answers is not in the document]
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:3555]: done

File: siliconcompiler/schema/schema_cfg.py (hunk 12 above)
Comment (aolofsson) [reply siliconcompiler/schema/schema_cfg.py:3508]: thanks, fixed

**Pull request merged**